st2common/st2common/util/actionalias_helpstring.py

```diff
@@ -0,0 +1,73 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from st2common.util.actionalias_matching import normalise_alias_format_string
+
+
+__all__ = [
+    'generate_helpstring_list'
+]
+
+
+def generate_helpstring_list(aliases, filtering="", pack="", limit=0, offset=0):
+    """
+    List help strings from a collection of alias objects.
+
+    :param aliases: The list of aliases
+    :type  aliases: ``list`` of :class:`st2common.models.api.action.ActionAliasAPI`
+    :param filtering: A search pattern.
+    :type  filtering: ``string``
+    :param pack: Name of a pack
+    :type  pack: ``string``
+    :param limit: The number of help strings to return in the list.
+    :type  limit: ``integer``
+    :param offset: The offset in the list to start return help strings.
+    :type  limit: ``integer``
+
+    :return: A list of aliases help strings.
+    :rtype: ``list`` of ``list``
+    """
+    matches = {}
+    cnt = 0
+    for alias in aliases:
+        # Skip disable aliases.
+        if not alias.enabled:
+            continue
+        # Skip packs which don't explicitly match the requested pack.
+        if pack != alias.pack and pack != "":
+            continue
+        for format_ in alias.formats:
+            display, _ = normalise_alias_format_string(format_)
+            if display:
+                # Skip over help strings until the requested offset is reached.
+                if cnt < offset:
```

Comment:
Sorry, I didn't make the intention of offset very clear in my initial post.  I'll try to explain it better here.  This feature was added after discussing the idea of a help api with @emedvedev.

The intention of offset was to allow paging through a large list of help strings.  If an st2 installation had a large number of packs and a client wanted to "scroll" through the list of helpstrings,  a client could make consecutive calls using the offset/limit pair.  E.g. if there were 122 help strings and the display interface was large enough for 20 items, the first call would contain `offset:0, limit: 20`, the second `offset:20, limit:20` etc.

Unfortunately, applying offset at line 53 wouldn't produce the result mentioned above because at that point in the code, `alias.formats[offset:]` is being applied to a single action alias and not the entire list of help strings.